File: Quick Sort/C++/quicksort.cpp
#include <iostream>
#include <algorithm>

int split(int a[], int start_index, int end_index)
{
    int x = a[end_index];
    int i = start_index - 1;
    for (int j = start_index; j < end_index; j++)
    {
        if (a[j] <= x)
        {
            i++;
            std::swap(a[i], a[j]);//swap(a[i],a[j])
        }
    }
    std::swap(a[i+1], a[end_index]);//swap(a[i+1],a[end_index])

    return i;
}

void quicksort(int a[], int start_index, int end_index)
{
    if (start_index < end_index)
    {
        int mid_index = split(a, start_index, end_index);
                        quicksort(a, start_index, mid_index - 1);
                        quicksort(a, mid_index + 1, end_index);
    }
}

int main()
{
    int arr[8] = {2, 8, 7, 1, 3, 5, 6, 4};
    quicksort(arr, 0, 7);
    for (int i = 0; i < 8; i++)
    {
        std::cout << arr[i] << " ";
    }
    std::cout << "\n";
    return 0;
}

File: Quick Sort/quickSort.py
def get_input():
    """
    get input from user
    """
    input_str = input("Enter elements to be sorted: ")
    try:
        elements = [int(e) for e in input_str.split()] #make a list of integers from input string
    except ValueError:
        print("Please enter a list of integers only, seperated by a space!!")
    return elements

def partition(thelist, start_idx: int, end_idx: int):
    """
    partition list according to pivot and return index of pivot
    """
    pivot = thelist[end_idx] #select last element as the pivot
    idx = start_idx - 1

    for curr_idx in range(start_idx, end_idx):
        if thelist[curr_idx] <= pivot:
            idx += 1 #increment
            thelist[idx], thelist[curr_idx] = thelist[curr_idx], thelist[idx] #swapping

    thelist[idx+1], thelist[end_idx] = thelist[end_idx], thelist[idx+1] #swapping
    return idx+1 #returning pivot index

def quick_sort(thelist, start_idx: int, end_idx: int):
    """
    Quick Sort Algorithm
    """
    if len(thelist) == 0:
        print("Empty list!!")

    elif len(thelist) == 1:
        print("Only one element!!")

    elif start_idx < end_idx:
        pivot_idx = partition(thelist, start_idx, end_idx) #get pivot index
        quick_sort(thelist, start_idx, pivot_idx - 1) #apply algorithm for smaller list
        quick_sort(thelist, pivot_idx + 1, end_idx) #apply algorithm for smaller list

if __name__ == '__main__':
    input_list = get_input()
    quick_sort(input_list, 0, len(input_list) - 1)
    print(*input_list, sep = ", ")
